packages/compile-solidity/src/compilerSupplier/loadingStrategies/Docker.ts

```diff
@@ -0,0 +1,227 @@
+import axios from "axios";
+import axiosRetry from "axios-retry";
+import fs from "fs";
+import { execSync } from "child_process";
+import ora from "ora";
+import semver from "semver";
+import { Cache } from "../Cache";
+import { normalizeSolcVersion } from "../normalizeSolcVersion";
+import { NoVersionError, NoRequestError } from "../errors";
+import { asyncFirst, asyncFilter, asyncFork } from "iter-tools";
+
+export class Docker {
+  private config: {
+    spawn: {
+      maxBuffer: number;
+    };
+    dockerTagsUrl: string;
+    version: string;
+  };
+  private cache: Cache;
+
+  constructor(options) {
+    const defaultConfig = {
+      dockerTagsUrl:
+        "https://registry.hub.docker.com/v2/repositories/ethereum/solc/tags/"
+    };
+    this.config = Object.assign({}, defaultConfig, options);
+
+    this.cache = new Cache();
+  }
+
+  async load() {
+    // Set a sensible limit for maxBuffer
+    // See https://github.com/nodejs/node/pull/23027
+    let maxBuffer = 1024 * 1024 * 100;
+    if (this.config.spawn && this.config.spawn.maxBuffer) {
+      maxBuffer = this.config.spawn.maxBuffer;
+    }
+
+    const versionString = await this.validateAndGetSolcVersion();
+    const command =
+      "docker run --rm -i ethereum/solc:" +
+      this.config.version +
+      " --standard-json";
+
+    try {
+      return {
+        compile: options =>
+          String(execSync(command, { input: options, maxBuffer })),
+        version: () => versionString
+      };
+    } catch (error) {
+      if (error.message === "No matching version found") {
+        throw new NoVersionError(versionString);
+      }
+      throw error;
+    }
+  }
+
+  /**
+   * Fetch list of solc versions available as Docker images.
+   *
+   * This returns a promise for an object with three fields:
+   *   { latestRelease, releases, prereleases }
+   * NOTE that `releases` and `prereleases` in this object are both
+   * AsyncIterableIterators (thus, use only `for await (const ...)` to consume)
+   */
+  async list() {
+    const allTags = this.streamAllDockerTags();
+
+    // split stream of all tags into separate releases and prereleases streams
+    const isRelease = name => !!semver.valid(name);
+    const isPrerelease = name => name.match(/nightly/);
+    const [allTagsA, allTagsB] = asyncFork(allTags);
+
+    // construct prereleases stream
+    const prereleases = asyncFilter(isPrerelease, allTagsB);
+
+    // construct releases stream and immediately fork so as to allow consuming
+    // the first value in the stream safely
+    const [releases, forkedReleases] = asyncFork(
+      asyncFilter(isRelease, allTagsA)
+    );
+
+    // grab the latest release from the forked releases stream;
+    // coerce semver to remove possible `-alpine` suffix used by this repo
+    const latestRelease = semver.coerce(
+      await asyncFirst(forkedReleases)
+    )?.version;
+
+    return {
+      prereleases,
+      releases,
```

Comment:
@haltman-at writes [here](https://github.com/trufflesuite/truffle/pull/4222#discussion_r687297120) how it's confusing that these are AsyncIterables, which warrants better commenting.

Reply:
Noted a comment [here](https://github.com/trufflesuite/truffle/pull/4262#diff-94e2cd9540d4f8bd22a774f9e893ef938ff57e71e4b18d2a4f2c51f79f48a83fR73-R82)